### PR TITLE
Implement most of RUTv2 features.

### PR DESCRIPTION
**Pull request by yuchenshi**

### CODE REVIEW

File: packages/rules-unit-testing/src/impl/test_environment.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    this.app = undefined;
+  }
+
+  firestore(settings?: FirestoreSettings): Firestore {
+    assertEmulatorRunning(this.emulators, 'firestore');
+    const firestoreCompat = this.getApp().firestore!();
+    if (settings) {
+      firestoreCompat.settings(settings);
+    }
+    const firestore = firestoreCompat as unknown as Firestore;
```
Comment (sam-gc) [inline packages/rules-unit-testing/src/impl/test_environment.ts:179]: This shouldn't be necessary :/

File: packages/rules-unit-testing/src/util.ts [hunk 2]
```diff
@@ -70,7 +70,34 @@ export async function withFunctionTriggersDisabled<TResult>(
...
 export function assertFails(pr: Promise<any>): Promise<any> {
-  throw new Error('unimplemented');
+  return pr.then(
+    () => {
+      return Promise.reject(
+        new Error('Expected request to fail, but it succeeded.')
+      );
+    },
+    (err: any) => {
+      const errCode = (err && err.code && err.code.toLowerCase()) || '';
```
Comment (sam-gc) [inline packages/rules-unit-testing/src/util.ts:80]: Can just do `err?.code?.toLowerCase() || ''`

File: packages/rules-unit-testing/src/util.ts [hunk 3]
```diff
@@ -70,7 +70,34 @@ export async function withFunctionTriggersDisabled<TResult>(
...
+      const errMessage =
+        (err && err.message && err.message.toLowerCase()) || '';
```
Comment (sam-gc) [inline packages/rules-unit-testing/src/util.ts:82]: Same here

**Review by sam-gc — APPROVED** [review]

File: packages/rules-unit-testing/src/impl/test_environment.ts (hunk 1 above)
Comment (yuchenshi) [reply packages/rules-unit-testing/src/impl/test_environment.ts]: Fixed after rebasing v9 changes in

File: packages/rules-unit-testing/src/util.ts (hunk 2 above)
Comment (yuchenshi) [reply packages/rules-unit-testing/src/util.ts]: Done

File: packages/rules-unit-testing/src/util.ts (hunk 3 above)
Comment (yuchenshi) [reply packages/rules-unit-testing/src/util.ts]: Done

**Pull request merged**